server/tso/allocator_manager.go

```diff
@@ -656,3 +679,83 @@ func (am *AllocatorManager) GetLocalAllocatorLeaders() (map[string]*pdpb.Member,
 	}
 	return localAllocatorLeaderMember, nil
 }
+
+func (am *AllocatorManager) getOrCreateGRPCConn(ctx context.Context, addr string) (*grpc.ClientConn, error) {
+	conn, ok := am.getGRPCConn(addr)
+	if ok {
+		return conn, nil
+	}
+	tlsCfg, err := am.securityConfig.ToTLSConfig()
+	if err != nil {
+		return nil, err
+	}
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, dialTimeout)
+	defer cancel()
+	cc, err := grpcutil.GetClientConn(ctxWithTimeout, addr, tlsCfg)
+	if err != nil {
+		return nil, err
+	}
+	am.setGRPCConn(cc, addr)
+	return cc, nil
+}
+
+func (am *AllocatorManager) isLeaderAwareOfDCLocation(ctx context.Context, dcLocation string) (bool, error) {
+	dcLocations, err := am.getLeaderDCLocations(ctx)
+	if err != nil {
+		return false, err
+	}
+	for _, dc := range dcLocations {
+		if dcLocation == dc {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (am *AllocatorManager) getLeaderDCLocations(ctx context.Context) ([]string, error) {
+	dcLocations := make([]string, 0)
+	if am.member.IsLeader() {
+		for dcLocation := range am.GetClusterDCLocations() {
+			dcLocations = append(dcLocations, dcLocation)
+		}
+		return dcLocations, nil
+	}
+
+	leaderAddrs := am.member.GetLeader().GetClientUrls()
+	if leaderAddrs == nil || len(leaderAddrs) < 1 {
+		return nil, fmt.Errorf("failed to get leader client url")
+	}
+	conn, err := am.getOrCreateGRPCConn(ctx, leaderAddrs[0])
+	if err != nil {
+		return nil, err
+	}
+	getCtx, cancel := context.WithTimeout(ctx, rpcTimeout)
+	defer cancel()
+	resp, err := pdpb.NewPDClient(conn).GetDCLocations(getCtx, &pdpb.GetDCLocationsRequest{
+		Header: &pdpb.RequestHeader{
+			SenderId: am.member.GetLeader().GetMemberId(),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.GetDcLocations(), nil
+}
+
+func (am *AllocatorManager) getGRPCConn(addr string) (*grpc.ClientConn, bool) {
+	am.localAllocatorConn.RLock()
+	defer am.localAllocatorConn.RUnlock()
+	conn, ok := am.localAllocatorConn.clientConns[addr]
+	return conn, ok
+}
+
+func (am *AllocatorManager) setGRPCConn(newConn *grpc.ClientConn, addr string) {
+	am.localAllocatorConn.Lock()
+	defer am.localAllocatorConn.Unlock()
+	if _, ok := am.localAllocatorConn.clientConns[addr]; ok {
+		newConn.Close()
+		log.Debug("use old connection", zap.String("target", newConn.Target()), zap.String("state", newConn.GetState().String()))
+		return
```

Comment:
In this case, we should return the old connection. Otherwise, `getOrCreateGRPCConn` will make a closed connection return as the result.